sycl/include/sycl/handler.hpp

```diff
@@ -970,6 +998,30 @@ class __SYCL_EXPORT handler {
     constexpr bool UsesRootSync = PropertiesT::template has_property<
         sycl::ext::oneapi::experimental::use_root_sync_key>();
     setKernelIsCooperative(UsesRootSync);
```

Comment:
I'm surprised that we don't appear to be modifying this (or something like it) based on the progress properties.  This might be my fault for not documenting things clearly in the [implementation notes](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc#implementation-notes) section of the specification.

Although the table there says "concurrent" for "GPU & Level Zero", that's because the table was intended to represent the _maximum_ capability, not the default.  By default, I think work-group progress is "parallel" -- if the user requested "concurrent" then we need a cooperative launch.  I'm honestly not sure if a cooperative launch is required for "concurrent" at sub-group scope.

Enabling cooperative launch for "concurrent" at work-group or sub-group scope might be a good idea for now (to make sure that users get the guarantee they asked for) but I think there needs to be some follow-up design work to decide how to express this at a UR level.  Different backends may behave differently, and may not require a cooperative launch in the cases that Level Zero does.

Reply:
> I'm surprised that we don't appear to be modifying this (or something like it) based on the progress properties. This might be my fault for not documenting things clearly in the [implementation notes](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc#implementation-notes) section of the specification.
> 
> Although the table there says "concurrent" for "GPU & Level Zero", that's because the table was intended to represent the _maximum_ capability, not the default. By default, I think work-group progress is "parallel" -- if the user requested "concurrent" then we need a cooperative launch. I'm honestly not sure if a cooperative launch is required for "concurrent" at sub-group scope.
> 
> Enabling cooperative launch for "concurrent" at work-group or sub-group scope might be a good idea for now (to make sure that users get the guarantee they asked for) but I think there needs to be some follow-up design work to decide how to express this at a UR level. Different backends may behave differently, and may not require a cooperative launch in the cases that Level Zero does.

Thanks for the knowledge shared, I'll make the change and add a TODO item as a comment asking to clarify this at a later stage.

Reply:
@JackAKirk - Does the same logic here apply to the CUDA backend?  I think it does, but I just want to double check.  If the same logic applies, would enabling a cooperative kernel call for CUDA here work as expected, or should the CUDA backend work be left for a future PR?

Reply:
> @JackAKirk - Does the same logic here apply to the CUDA backend? I think it does, but I just want to double check. If the same logic applies, would enabling a cooperative kernel call for CUDA here work as expected, or should the CUDA backend work be left for a future PR?

@JackAKirk ping

Reply:
Sorry I missed the first ping.

At the subgroup level I read that nvidia post volta satisfies the C++ forward progress guarantees which seems to coincide with your extension document. The best resource I know of explaining this is https://developer.nvidia.com/blog/inside-volta/ 
I don't have access to it but this doc might provide further details: https://developer.nvidia.com/blog/inside-volta/
Also see https://on-demand.gputechconf.com/gtc/2018/presentation/s81006-volta-architecture-and-performance-optimization.pdf

Now wrt the question I think you are asking: 
- how does nvidia `cudaLaunchCooperativeKernel` "weaken" these guarantees as defined in your extension?

 I don't know the answer to this and I don't see mention of it in Nvidia docs etc.
I don't know the details of C++ iso forward progress guarantees but I think I get the basic idea; the other thing that is relevant to note is that `cudaLaunchCooperativeKernel` is not the only way that these guarantees might be affected from what I understand. `cudaLaunchKernelExC`

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1ge236ecdbbaf7cf47a806bba71c1d03c4

 is a newer api (requires sm90 "hopper") and allows access of distributed shared memory between different thread blocks, as well as interthread block atomics to control such distributed shared memory that requires their ptx 7.8 memory model:

https://dl.acm.org/doi/pdf/10.1145/3470496.3533045

It is as far as I can see completely separate from `cudaLaunchCooperativeKernel` and provides non overlapping functionality. The impression I get is that going forward this will be a more important means of inter-thread block interaction than `cudaLaunchCooperativeKernel` (I note also that the multi device version of this api has been deprecated).

Now I may have misunderstood some things but I guess that both `cudaLaunchCooperativeKernel` and `cudaLaunchKernelExC` (with associated cluster_group functionality) https://github.com/intel/llvm/pull/13594 might affect your guarantees, and possibly in different ways.

Hope this is in some way useful.

Reply:
> Sorry I missed the first ping.
> 
> At the subgroup level I read that nvidia post volta satisfies the C++ forward progress guarantees which seems to coincide with your extension document. The best resource I know of explaining this is https://developer.nvidia.com/blog/inside-volta/ I don't have access to it but this doc might provide further details: https://developer.nvidia.com/blog/inside-volta/ Also see https://on-demand.gputechconf.com/gtc/2018/presentation/s81006-volta-architecture-and-performance-optimization.pdf
> 
> Now wrt the question I think you are asking:
> 
> * how does nvidia `cudaLaunchCooperativeKernel` "weaken" these guarantees as defined in your extension?
> 
> I don't know the answer to this and I don't see mention of it in Nvidia docs etc. I don't know the details of C++ iso forward progress guarantees but I think I get the basic idea; the other thing that is relevant to note is that `cudaLaunchCooperativeKernel` is not the only way that these guarantees might be affected from what I understand. `cudaLaunchKernelExC`
> 
> https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1ge236ecdbbaf7cf47a806bba71c1d03c4
> 
> is a newer api (requires sm90 "hopper") and allows access of distributed shared memory between different thread blocks, as well as interthread block atomics to control such distributed shared memory that requires their ptx 7.8 memory model:
> 
> https://dl.acm.org/doi/pdf/10.1145/3470496.3533045
> 
> It is as far as I can see completely separate from `cudaLaunchCooperativeKernel` and provides non overlapping functionality. The impression I get is that going forward this will be a more important means of inter-thread block interaction than `cudaLaunchCooperativeKernel` (I note also that the multi device version of this api has been deprecated).
> 
> Now I may have misunderstood some things but I guess that both `cudaLaunchCooperativeKernel` and `cudaLaunchKernelExC` (with associated cluster_group functionality) #13594 might affect your guarantees, and possibly in different ways.
> 
> Hope this is in some way useful.

Thank you @JackAKirk !
@Pennycook Should I leave the cuda changes for a future PR?

Reply:
Thanks, @JackAKirk .  This is useful -- I didn't know about the existence of `cudaLaunchKernelExc`.

@lbushi25 - My recommendation here for now is to change the implementation to:

```c++
    if (Guarantee == forward_progress::concurrent)
      setKernelIsCooperative(true);
```

...so that it's required for all backends.  It sounds like the UR will then need to decide between `cudaLaunchCooperativeKernel`, `cudaLaunchKernelExC` for CUDA.

I think moving the complexity out of the headers and into UR is the right thing to do here.